### PR TITLE
Luxai2022 - beta update 1.0.1

### PR DESCRIPTION
**Pull request by StoneT2000**
- Soft limit on unit control (1 power / 10 power cost for light and heavy comms)
- Fixed factory peaceable overlapping halfway line
- better map gen with more distance between nearest resources and smaller clusters
@bovard

### CODE REVIEW

File: kaggle_environments/envs/lux_ai_2022/luxai2022/env.py [hunk 1]
```diff
@@ -482,12 +482,18 @@ def step(self, actions):
                         if "factory" in unit_id:
                             self.state.factories[agent][unit_id].action_queue.append(format_factory_action(action))
                         elif "unit" in unit_id:
+                            unit = self.state.units[agent][unit_id]
+                            # if unit does not have more than 5 power, we skip updating the action queue and print warning
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/env.py:486]: `UNIT_ACTION_QUEUE_POWER_COST`

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/config.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            RUBBLE=dict(LIGHT=1, HEAVY=10),
+            TIME_RANGE=[1, 5]
+        ),
+        COLD_SNAP=dict(
+            # power multiplier required per robot action. 2 -> requires 2x as much power to execute the same action
+            POWER_CONSUMPTION=2,
+            TIME_RANGE=[10, 30]
+        ),
+        DUST_STORM=dict(
+            # power multiplier required per robot action. .5 -> requires .5x as much power to execute the same action
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/config.py:127]: I thought this was gain .5x as much power per turn (dust makes it harder to gain power?)

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+    def build_heavy_metal_cost(self, game_state):
+        unit_cfg = self.env_cfg.ROBOTS["HEAVY"]
+        return unit_cfg.METAL_COST
+    def build_heavy_power_cost(self, game_state):
+        current_weather = game_state.weather_schedule[game_state.real_env_steps]
+        weather_cfg = get_weather_config(current_weather, self.env_cfg)
+        unit_cfg = self.env_cfg.ROBOTS["HEAVY"]
+        return math.ceil(unit_cfg.POWER_COST * weather_cfg["power_loss_factor"])
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py:33]: I thought we were going to have factories immune to weather (due to their alternative power source?)

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/config.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    MAX_FACTORIES: int = 5
+    CYCLE_LENGTH: int = 50
+    DAY_LENGTH: int = 30
+    UNIT_ACTION_QUEUE_SIZE: int = 10 # when set to 1, then no action queue is used
+    UNIT_ACTION_QUEUE_POWER_COST: Dict[str, int] = dataclasses.field(default_factory=lambda: dict(
+        LIGHT=1,
+        HEAVY=10
+    ))
+    # DEPRECATED
+    # UNITS_CONTROLLED: int = 20 # when set to -1, all units can be controlled at once
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/config.py:58]: can just remove this?

File: kaggle_environments/envs/lux_ai_2022/luxai2022/config.py [hunk 5]
```diff
@@ -60,16 +69,23 @@ class EnvConfig:
...
 
-    #### Factores ####
-    FACTORY_PROCESSING_RATE_WATER: int = 50
-    ICE_WATER_RATIO: int = 1
+    #### Factories ####
+    FACTORY_PROCESSING_RATE_WATER: int = 100
+    ICE_WATER_RATIO: int = 5
     FACTORY_PROCESSING_RATE_METAL: int = 50
-    ORE_METAL_RATIO: int = 1
+    ORE_METAL_RATIO: int = 5
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/config.py:82]: 👍

File: kaggle_environments/envs/lux_ai_2022/luxai2022/map_generator/visualize_samples.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+from math import prod
+from pathlib import Path
+from luxai2022.env import LuxAI2022
+import pygame
+import numpy as np
+if __name__ == "__main__":
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/map_generator/visualize_samples.py:6]: nice!

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/unit.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        factory_there = board.factory_occupancy_map[target_pos[1], target_pos[0]]
+        if factory_there != self.team_id and factory_there != -1:
+            # print("Warning, tried to get move cost for going onto a opposition factory", file=sys.stderr)
+            return None
+        rubble_at_target = board.rubble[target_pos[1]][target_pos[0]]
+        
+        current_weather = game_state.weather_schedule[game_state.real_env_steps]
+        weather_cfg = get_weather_config(current_weather, self.env_cfg)
+        return math.ceil((self.unit_cfg.MOVE_COST + self.unit_cfg.RUBBLE_MOVEMENT_COST * rubble_at_target) * weather_cfg["power_loss_factor"])
+    def can_move(self, game_state, direction):
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/unit.py:51]: scrap this or need to factor the cost to send the command into it

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/unit.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    def pickup(self, pickup_resource, pickup_amount, repeat=True):
+        assert pickup_resource < 5 and pickup_resource >= 0
+        return np.array([2, 0, pickup_resource, pickup_amount, 1 if repeat else 0])
+    
+    def dig_cost(self, game_state):
+        current_weather = game_state.weather_schedule[game_state.real_env_steps]
+        weather_cfg = get_weather_config(current_weather, self.env_cfg)
+        return math.ceil(self.unit_cfg.DIG_COST * weather_cfg["power_loss_factor"])
+    def can_dig(self, game_state):
+        return self.power >= self.dig_cost(game_state)
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/unit.py:77]: same with this

**Review by bovard — CHANGES_REQUESTED** [review]

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/config.py (hunk 2 above)
Comment (StoneT2000) [reply kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/config.py:127]: oops good catch

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py (hunk 3 above)
Comment (StoneT2000) [reply kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py:33]: Oh I figured cold snaps applied to all machinery? All machines need to warm up more? Also one less specific thing to memorize in addition to I think cold snap applying to everything being intuitive.

**Comment by StoneT2000** [comment]
Pushed a bug fix for marsquakes not working. Should be good now

File: kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py (hunk 3 above)
Comment (StoneT2000) [reply kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py:33]: Maybe you were thinking that in dust storms factories don't gain extra power (which is the case).

Comment (bovard) [reply kaggle_environments/envs/lux_ai_2022/test_agents/python/lux/factory.py:33]: yeah that's it!

**Pull request merged**